### PR TITLE
Add special treatment for Java's init block #108

### LINKED ISSUE #108: JAVA Initialization block should be counted for complexity and functions metrics
Java initialization blocks contain logic, which would be executed when an instance is created or the class is loaded (static initialization block). They are similar to init block in Kotlin. Since we already count init block in Kotlin as complexity and functions, we should the same for Java initialization block.

### PR DESCRIPTION
**Pull request by mylinhdao**
Closes #108

### CODE REVIEW

File: src/parser/metrics/Complexity.ts [hunk 1]
```diff
@@ -49,7 +51,17 @@ export class Complexity implements Metric {
...
+            case Languages.Java:
+                queryBuilder.setStatements(
+                    this.mccStatementsSuperSet.concat(
+                        new SimpleQueryStatement("(class_body (block)) @initBlock")
+                    )
+                );
+                break;
+            default:
+                queryBuilder.setStatements(this.mccStatementsSuperSet);
+        }
```
Comment (ResistantBear) [inline src/parser/metrics/Complexity.ts:64]: @ce-bo suggested that we should not use switch-cases in a similar situation until we have more than one case to differentiate (see https://github.com/MaibornWolff/metric-gardener/pull/74#discussion_r1479348223), so we should also refactor this to an if-statement here.

File: src/parser/metrics/Functions.ts [hunk 2]
```diff
@@ -19,7 +24,17 @@ export class Functions implements Metric {
...
+            case Languages.Java:
+                queryBuilder.setStatements(
+                    this.statementsSuperSet.concat(
+                        new SimpleQueryStatement("(class_body (block)) @initBlock")
+                    )
+                );
+                break;
+            default:
+                queryBuilder.setStatements(this.statementsSuperSet);
+        }
```
Comment (ResistantBear) [inline src/parser/metrics/Functions.ts:37]: The same applies here. Refactor to an if-statement.

**Review by ResistantBear — APPROVED** [review]
Seems like you are right that we actually need a special treatment here to find the non-static initialization blocks in Java. How unfortunate.

Looks fine, but please take care of the remarks.

File: src/parser/metrics/Functions.ts (hunk 2 above)
Comment (mylinhdao) [reply src/parser/metrics/Functions.ts:36]: i changed to if else as suggested

File: src/parser/metrics/Complexity.ts [hunk 3]
```diff
@@ -49,7 +50,16 @@ export class Complexity implements Metric {
...
+        if (parseFile.language === Languages.Java) {
+            //add query for instance init block in Java
+            queryBuilder.setStatements(
+                this.mccStatementsSuperSet.concat(
+                    new SimpleQueryStatement("(class_body (block)) @initBlock")
+                )
+            );
+        } else {
+            queryBuilder.setStatements(this.mccStatementsSuperSet);
+        }
```
Comment (ce-bo) [inline src/parser/metrics/Complexity.ts:62]: Isn't it possible like this to shorten the code?

```suggestion
        if (parseFile.language === Languages.Java) {
            //add query for instance init block in Java
            this.mccStatementsSuperSet.concat(
                new SimpleQueryStatement("(class_body (block)) @initBlock")
            );
        }
        queryBuilder.setStatements(this.mccStatementsSuperSet);
```

**Review by ce-bo — APPROVED** [review]
In general LGTM :-) Thank you!

Unfortunately I was not aware that we need to add a special language handling for java init blocks. I would like to avoid that in general and to keep metric-gardener as simple and stupid as possible. Even if that means that some new or special language features are not supported (yet).

Let's merge this implementation because it does not add much complexity but let's discuss such cases for upcoming implementations.

File: src/parser/metrics/Complexity.ts (hunk 3 above)
Comment (ResistantBear) [reply src/parser/metrics/Complexity.ts:62]: No. `concat` actually returns a new array object that copies all elements. And that is on purpose, as mutating the class field (in this case `this.mccStatementsSuperSet`) like you seem to suggest here, e.g. by using `push` which mutates the existing array, would be a side effect that also affects further runs on other files. That was the exact issue that caused #65 and should be avoided therefore.

**Pull request merged**